specification.md

```diff
@@ -0,0 +1,257 @@
+#   GOLDFLAGS: Extra linking flags passed to 'go' when building.
+#   GOGCFLAGS: Additional go compile flags passed to 'go' when building.
+#
+# Example:
+#   make
+#   make all or make build
+#   make build WHAT=cmd/server GOFLAGS=-v
+#   make all GOGCFLAGS="-N -l"
+#     Note: Use the -N -l options to disable compiler optimizations an inlining.
+#           Using these build options allows you to subsequently use source
+#           debugging tools like delve.
+```
+
+### Color log
+
+The Makefile SHOULD print colorized log when `$COLOR_LOG == true`
+
+### Version
+
+The valid version MUST be able to describe the current status of git tree.
+
+A complete version information contains:
+
+-   a git tag versioned by [Semantic Versioning 2.0.0](http://semver.org/spec/v2.0.0.html) (semver). e.g. v1.0.0 v0.1.0-alpha.1
+-   commits number after latest git tag and latest commit hash
+-   git tree status: 
+    -   `clean` indicates no changes since the git commit id.
+    -   `dirty` indicates source code changes after the git commit id
+    -   `archive` indicates the tree was produced by `git archive`
+
+#### Use Case
+
+| version                 | docker tag              | description                              |
+| ----------------------- | ----------------------- | ---------------------------------------- |
+| v0.0.1                  | v0.0.1                  | a clean git tree with the latest tag `v0.0.1` |
+| v0.0.1-dirty            | v0.0.1-dirty            | a dirty git tree with changes not staged |
+| v0.0.1-2+1b4531e1acf800 | v0.0.1-2-1b4531e1acf800 | a clean git tree with a tag ` v0.0.1`.<br />There are two new commits after the tag and the latest commit is `1b4531e1acf800` |
+
+If your git repo contains changes not staged, a `-dirty` suffix MUST always be appended to final version.
+
+#### Options
+
+| Name    | Usage                     | Type   | Default | Severity |
+| ------- | ------------------------- | ------ | ------- | -------- |
+| VERSION | used to replace `git tag` | string |         | REQUIRED |
+
+#### Example
+
+```bash
+VERSION=v0.0.1 make all
+```
+
+### Compile
+
+The Makefile MUST support the following targets for compiling.
+
+```bash
+make all # equal to make build
+make build
+make build-local
+make build-in-container
+```
+
+`make all` MUST be equal to `make build`
+
+`make build` means to compile you project. The Makefile MUST support the ability to compile the project on localhost and in docker container.
+
+ For convenience, the Makefile SHOULD support the ability to build targets directories directly.
+
+```bash
+make cmd/controller # equal to make build WHAT=cmd/controller
+```
+
+#### Options
+
+| Name        | Usage                                    | Type   | Default | Severity |
+| ----------- | ---------------------------------------- | ------ | ------- | -------- |
+| LOCAL_BUILD | If set true, project will be built on local machine, otherwise, built in docker. | string | false   | REQUIRED |
```

Comment:
nit: set to true

Reply:
why do I need `LOCAL_BUILD` if I have `make build-local` and `make build-in-docker`?

Reply:
`LOCAL_BUILD` is useful when you want to use `make cmd/controller` to build your app.
The default behavior of build will compile the app on localhost. If you want to compile the app in docker, there are three ways achieve it
```bash
make cmd/controller LOCAL_BUILD=false
make build-in-container WHAT=cmd/controller
make build WHAT=cmd/controller WHAT=cmd/controller
```
Somebody may prefer the first solution.

Reply:
What's is this for?  Typo?

```
make build WHAT=cmd/controller WHAT=cmd/controller
```